### PR TITLE
Add PSO training mode to ScalarQuantizationTrainer

### PR DESCRIPTION
**Pull request by HuXin0817**
Traditional training methods or TRUNC's training approach are unable to adapt to complex data distributions. We introduce machine learning algorithms to train the error parameters of SQ, and testing results demonstrate that the mean squared error achieves below 95% of the original error.

Why Choose the PSO Algorithm?

In my latest paper, I compared the PSO algorithm with an EM-like boundary training algorithm used in the Faiss library. I found that, under the same training time, the MSE of the PSO algorithm was lower than that of Faiss on the SIFT, Word2Vec, GIST, MSong, ImageNet, and Deep datasets. This is attributed to the swarm intelligence of PSO, which helps avoid local optima.

### CODE REVIEW

File: tests/test_scalar_quantization_trainer.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
+
```
Comment (LHT129) [inline tests/test_scalar_quantization_trainer.cpp]: you can rename the file and move it to the same directory of "scalar_quantization_trainer.cpp" just like other unittest files

Comment (HuXin0817) [reply tests/test_scalar_quantization_trainer.cpp]: Thanks reminding, I will fix it soon.

Comment (HuXin0817) [reply tests/test_scalar_quantization_trainer.cpp]: I have modified my commit.

**Comment by inabao** [comment]
are there any experimental results?

**Review by ShawnShawnYou — CHANGES_REQUESTED** [review]
add END2END performance comparison by eval_performance on datasets like GIST, SIFT...

File: src/quantization/scalar_quantization/scalar_quantization_trainer.h [hunk 2]
```diff
@@ -37,7 +38,7 @@ class ScalarQuantizationTrainer {
           float* upper_bound,
           float* lower_bound,
           bool need_normalize = false,
-          SQTrainMode mode = SQTrainMode::TRUNC_BOUND);
+          SQTrainMode mode = SQTrainMode::PSO);  // suggest to use PSO
```
Comment (ShawnShawnYou) [inline src/quantization/scalar_quantization/scalar_quantization_trainer.h]: add parameters in SQ to choose PSO instead of modifying the default mode

File: src/quantization/scalar_quantization/scalar_quantization_trainer.cpp [hunk 3]
```diff
@@ -39,6 +39,137 @@ ScalarQuantizationTrainer::Train(const float* data,
...
+            float curr_loss = loss(particle.center, particle.width);
+            particle.min_loss = curr_loss;
+            if (curr_loss < global_min_loss) {
+                global_min_loss = curr_loss;
+                global_best_center = particle.center;
+                global_best_width = particle.width;
+            }
+        }
+
+        for (size_t iter = 0; iter < max_iter; ++iter) {
```
Comment (ShawnShawnYou) [inline src/quantization/scalar_quantization/scalar_quantization_trainer.cpp:163]: It is suggested to split this long function into smaller functions and add unit tests (UT) to cover each one of them

File: src/quantization/scalar_quantization/scalar_quantization_trainer_test.cpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    // CLASSIC
+    trainer.Train(data.data(), data.size(), &lower_c, &upper_c, vsag::SQTrainMode::CLASSIC);
+    float mse_classic = compute_mse(data, lower_c, upper_c, bits);
+
+    // TRUNC_BOUND
+    trainer.Train(data.data(), data.size(), &lower_t, &upper_t, vsag::SQTrainMode::TRUNC_BOUND);
+    float mse_trunc = compute_mse(data, lower_t, upper_t, bits);
+
+    // PSO
+    trainer.Train(data.data(), data.size(), &lower_p, &upper_p, vsag::SQTrainMode::PSO);
```
Comment (ShawnShawnYou) [inline src/quantization/scalar_quantization/scalar_quantization_trainer_test.cpp]: add tests to ensure that lower_p < upper_p across different datasets

Comment (HuXin0817) [reply src/quantization/scalar_quantization/scalar_quantization_trainer_test.cpp]: I will fix it soon.

**Comment by HuXin0817** [comment]
> Add END2END performance comparison using `eval_performance` on datasets like GIST, SIFT...

The specific code I have placed at [https://github.com/HuXin0817/POUQ/blob/deverlop/libpouq/optimizer.hpp](https://github.com/HuXin0817/POUQ/blob/deverlop/libpouq/optimizer.hpp). You can pull and test it.

File: src/quantization/scalar_quantization/scalar_quantization_trainer_test.cpp (hunk 4 above)
Comment (HuXin0817) [reply src/quantization/scalar_quantization/scalar_quantization_trainer_test.cpp]: done.

File: src/quantization/scalar_quantization/scalar_quantization_trainer.h (hunk 2 above)
Comment (HuXin0817) [reply src/quantization/scalar_quantization/scalar_quantization_trainer.h]: done.

File: src/quantization/scalar_quantization/scalar_quantization_trainer.cpp (hunk 3 above)
Comment (HuXin0817) [reply src/quantization/scalar_quantization/scalar_quantization_trainer.cpp:163]: It would be difficult to split it up, so I implemented a `pso_train_impl` function that can pass training parameters to facilitate possible testing later.
